src/Symfony/Component/Form/ChoiceList/ChoiceList.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\ChoiceList;
+
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceAttr;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceFieldName;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceLabel;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceLoader;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceValue;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\GroupBy;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\PreferredChoice;
+use Symfony\Component\Form\ChoiceList\Loader\CallbackChoiceLoader;
+use Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface;
+use Symfony\Component\Form\FormTypeExtensionInterface;
+use Symfony\Component\Form\FormTypeInterface;
+
+/**
+ * A set of convenient static methods to create cacheable choice list options.
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+final class ChoiceList
```

Comment:
I'm not sure I see the point of this class. I'd suggest removing it and inlining the instantiations instead.

Reply:
It's the user land part of this feature, everything else is about optimising the internals, many advantages to me:

- easier to document/learn, only one class instead of many (I also add the `ChoiceFilter` in #35733 and we may add more in the future like a `ChoiceLabelAttr` I'got locally)
- better discovery since every factory method has its own specific phpdoc instead of the generic constructor in the abstract class
- better auto complete, there are a lot of classes using the `Choice` word already:
  <img width="633" alt="Screenshot 2020-02-17 at 7 08 15 PM" src="https://user-images.githubusercontent.com/10107633/74677559-ef5e7a00-51b8-11ea-8cad-ca709dd864bd.png">

  VS 

  <img width="699" alt="Screenshot 2020-02-17 at 7 06 07 PM" src="https://user-images.githubusercontent.com/10107633/74677540-e66da880-51b8-11ea-85f0-9e1fd444fcc9.png">

I agree we could inline in our own usage in this PR to save a few static calls, but I wouldn't either since the types are some kind of documentation and given the Blackfire profiles I got with the current implementation.

Reply:
You can see this class as a `ChoiceList` builder or configurator